### DISCUSSION TITLE
Unable to run firejail and proxychains at the same time

### DISCUSSION THREAD

**Discussion by bussw83** [netblue30/firejail#5243] [Q&A]
Hello, I am new with Linux and have been using proxychains with firefox for some time.
I decided to install firejail but when I try to run firefox with proxychains, it seems that firejail blocks it. The browser is opened but proxychains don’t seem to run properly.

 I am using Parrot which I think uses Debian.

This is the output when I start proxychains with firefox:

```console
$ proxychains firefox
ProxyChains-3.1 (http://proxychains.sf.net)
Reading profile /etc/firejail/firefox.profile
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Parent pid 8263, child pid 8266
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Warning: cleaning all supplementary groups
Child process initialized in 200.60 ms
Warning: an existing sandbox was detected. /usr/bin/firefox-esr will run without any additional sandboxing features

###!!! [Parent][RunMessage] Error: Channel closing: too late to send/recv, messages will be lost

Parent is shutting down, bye…
```

I am sorry if this is a question that is too basic, I moved to linux a few weeks ago.

Thank you in advance :)

**Comment by rusty-snake** [comment]
#1093 #2699 #1160 #4156

TL;DR: Either run `firejail proxichains /usr/bin/firefox` with a `firefox.local` that allows proxychains or add `env LD_PRELOAD=<path to libproxychains.so>` to you firefox.local.

**Reply by bussw83** [reply]
Hello,

Thank you so much for your response I first tried to white list following your directions. But this was the output:
$firejail proxychains /usr/bin/firefox
Reading profile /etc/firejail/default.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-programs.inc

** Note: you can use --noprofile to disable default.profile **

Parent pid 6652, child pid 6653
Child process initialized in 87.82 ms
ProxyChains-3.1 (http://proxychains.sf.net)
run_symlink: LD_PRELOAD is: '/usr/lib/x86_64-linux-gnu/libproxychains.so.3'
firefox-esr: run_symlink.c:81: run_symlink: Assertion `env_get("LD_PRELOAD") == NULL' failed.

Parent is shutting down, bye...

I added this lines to firefox.local
whitelist /usr/share/doc/libproxychains3
whitelist /usr/share/doc/libproxychains
whitelist /usr/share/doc/proxychains
###############################
I also tried adding env LD_PRELOAD=<path to libproxychains.so>
This was the output:

$proxychains firefox
ProxyChains-3.1 (http://proxychains.sf.net)
Reading profile /etc/firejail/firefox.profile
Reading profile /etc/firejail/firefox.local
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Parent pid 6672, child pid 6675
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Warning: cleaning all supplementary groups
Child process initialized in 370.60 ms
ERROR: ld.so: object '</home/sugarbae/libproxychains.so.3>' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
ERROR: ld.so: object '</home/sugarbae/libproxychains.so.3>' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
run_symlink: LD_PRELOAD is: '</home/sugarbae/libproxychains.so.3>'
firefox-esr: run_symlink.c:81: run_symlink: Assertion `env_get("LD_PRELOAD") == NULL' failed.

Parent is shutting down, bye...



################################
I also tried this from the posts you shared:
firejail --env="LD_PRELOAD=libproxychains.so.3" <program>

The output was:
$firejail --env="LD_PRELOAD=libproxychains.so.3" firefox
Reading profile /etc/firejail/firefox.profile
Reading profile /etc/firejail/firefox.local
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Parent pid 7042, child pid 7045
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Warning: cleaning all supplementary groups
Child process initialized in 412.48 ms
ERROR: ld.so: object '<lib/x86_64-linux-gnu/libproxychains.so.3>' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
run_symlink: LD_PRELOAD is: '<lib/x86_64-linux-gnu/libproxychains.so.3>'
firefox: run_symlink.c:81: run_symlink: Assertion `env_get("LD_PRELOAD") == NULL' failed.

Parent is shutting down, bye...

############################################
I also tried:

$firejail --profile=/etc/firejail/firefox.profile proxychains firefox
Reading profile /etc/firejail/firefox.profile
Reading profile /etc/firejail/firefox.local
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Parent pid 7135, child pid 7138
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Warning: cleaning all supplementary groups
Child process initialized in 371.42 ms
ERROR: ld.so: object '<lib/x86_64-linux-gnu/libproxychains.so.3>' from LD_PRELOAD cannot be preloaded (cannot open shared object file): ignored.
ProxyChains-3.1 (http://proxychains.sf.net)
run_symlink: LD_PRELOAD is: '/usr/lib/x86_64-linux-gnu/libproxychains.so.3'
firefox: run_symlink.c:81: run_symlink: Assertion `env_get("LD_PRELOAD") == NULL' failed.

Parent is shutting down, bye...

########################################
I tried
firejail --profile=start-tor-browser --private-bin=proxychains --private-etc=proxychains.conf proxychains

This was the output:

Reading profile /etc/firejail/start-tor-browser.profile
Reading profile /etc/firejail/start-tor-browser.desktop.profile
Reading profile /etc/firejail/torbrowser-launcher.profile
Reading profile /etc/firejail/allow-python2.inc
Reading profile /etc/firejail/allow-python3.inc
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/disable-xdg.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Parent pid 7411, child pid 7412
46 programs installed in 86.54 ms
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Warning: skipping asound.conf for private /etc
Warning: skipping crypto-policies for private /etc
Warning: skipping pki for private /etc
Private /etc installed in 36.89 ms
Private /usr/etc installed in 0.00 ms
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Child process initialized in 245.53 ms
ProxyChains-3.1 (http://proxychains.sf.net)
usage:
proxychains <prog> [args]

Parent is shutting down, bye...

##################################

I know I am doing something wrong but I am not sure about what it is :(

**Reply by rusty-snake** [reply]
Please use code-blocks and details-summary tags for readability.

> firejail proxychains /usr/bin/firefox
Reading profile /etc/firejail/default.profile

Ooops, it should be `firejail --profile=firefox proxychains /usr/bin/firefox`

or `firejail --profile=firefox env LD_PRELOAD=<path to lib> /usr/bin/firefox`.

> run_symlink: LD_PRELOAD

The LD_PRELOAD issue.

> whitelist /usr/share/doc/libproxychains3
whitelist /usr/share/doc/libproxychains
whitelist /usr/share/doc/proxychains

1. This is documentation
2. This is already whitelisted

> /home/sugarbae/libproxychains.so.3

How and where did you installed proxychains?

> firejail --profile=/etc/firejail/firefox.profile proxychains firefox

This does not work. You *must* call firefox by path.

>  firejail --profile=start-tor-browser --private-bin=proxychains --private-etc=proxychains.conf proxychains

:open_mouth:

**Reply by bussw83** [reply]
Hello my bad I will make sure I use the brackets from now on :)

Yes I tried all I saw here, probably was not the best idea lol :')

I installed proxychains from a tutorial online, I have removed today and installed proxychains4

now I am getting this:

 
```
`firejail --profile=firefox env LD_PRELOAD=</usr/lib/x86_64-linux-gnu/libproxychains.so.4> /usr/bin/firefox
Reading profile /etc/firejail/firefox.profile
Reading profile /etc/firejail/firefox.local
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Warning: noroot option is not available
Reading profile /etc/firejail/whitelist-usr-share-common.inc
Reading profile /etc/firejail/firefox-common.profile
Reading profile /etc/firejail/disable-common.inc
Reading profile /etc/firejail/disable-devel.inc
Reading profile /etc/firejail/disable-exec.inc
Reading profile /etc/firejail/disable-interpreters.inc
Reading profile /etc/firejail/disable-proc.inc
Reading profile /etc/firejail/disable-programs.inc
Reading profile /etc/firejail/whitelist-common.inc
Reading profile /etc/firejail/whitelist-run-common.inc
Reading profile /etc/firejail/whitelist-runuser-common.inc
Reading profile /etc/firejail/whitelist-var-common.inc
Warning: noroot option is not available
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Parent pid 7255, child pid 7258
Warning: An abstract unix socket for session D-BUS might still be available. Use --net or remove unix from --protocol set.
Seccomp list in: !chroot, check list: @default-keep, prelist: unknown,
Child process initialized in 377.47 ms

Parent is shutting down, bye...`
```


What do you think? Seems that the preload is not longer an issue but still not running proxychains, thanks again for all your help :(

**Reply by rusty-snake** [reply]
<code>firejail --profile=firefox env LD_PRELOAD=<strong>&lt;path to lib&gt;</strong> /usr/bin/firefox</code> the strong part is the place holder.

**Reply by bussw83** [reply]
Hi, do you mean that the path to the lib I used was wrong?

**Reply by rusty-snake** [reply]
Yes. If it is at `/usr/lib/x86_64-linux-gnu/libproxychains.so.4`, the command is `firejail --profile=firefox env LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libproxychains.so.4 /usr/bin/firefox`.

**Comment by osevan** [comment]
https://github.com/netblue30/firejail/discussions/5243#discussioncomment-3139207


When i run this proxychains works but apparmor lib doesnt invoked.


:-(

**Comment by osevan** [comment]
Maybe i found a bug?
